[core] Revive clang-format

AndrewShkrob opened the pull request:
Closes: #3661

I want to split this feature into 3 PRs:
1. This one - updated clang-format file
2. Reformat all files according to new clang-format
3. Add check to CI

What's done with clang-format:
* Removed common Google-style configs
* Similar configs for Cpp and ObjC moved to common section
* Sorted in ABC order

AndrewShkrob commented:
@vng This PR is about `clang-format` not `clang-tidy`

biodranik commented:
Viktor is asking about formatted files changes.

AndrewShkrob commented:
I assume almost all files will be reformatted because you were not following your clang-format rules for years 😄

biodranik commented:
It would be great to have a preview of _how badly_ will the code be reformatted. Maybe some rules can be relaxed, etc.

AndrewShkrob commented:
Nothing bad will happen. I'd say it will be even better than now.
I (almost) didn't change the config, so you can try locally right now in your IDE and you'll see the differences.


Here are some examples:
```cpp
// Before
using TFlushFn = std::function<void (RenderState const &, drape_ptr<RenderBucket> &&)>;
// After
using TFlushFn = std::function<void(RenderState const &, drape_ptr<RenderBucket> &&)>;



// Before
std::string const & SymbolsTexture::SymbolKey::GetSymbolName() const
{
  return m_symbolName;
}
// After
std::string const & SymbolsTexture::SymbolKey::GetSymbolName() const { return m_symbolName; }



// Before
m2::RectF StipplePenPacker::MapTextureCoords(m2::RectU const & pixelRect) const
{
  return { (pixelRect.minX() + 0.5f) / m_canvasSize.x,
           (pixelRect.minY() + 0.5f) / m_canvasSize.y,
           (pixelRect.maxX() - 0.5f) / m_canvasSize.x,
           (pixelRect.maxY() - 0.5f) / m_canvasSize.y };
}
// After
m2::RectF StipplePenPacker::MapTextureCoords(m2::RectU const & pixelRect) const
{
  return {(pixelRect.minX() + 0.5f) / m_canvasSize.x, (pixelRect.minY() + 0.5f) / m_canvasSize.y,
          (pixelRect.maxX() - 0.5f) / m_canvasSize.x, (pixelRect.maxY() - 0.5f) / m_canvasSize.y};
}



// Before
void SetAttribute(std::string const & attributeName, uint32_t bufferInd, uint32_t offset, uint32_t componentsCount);
// After
void SetAttribute(std::string const & attributeName, uint32_t bufferInd, uint32_t offset,
                  uint32_t componentsCount);

// Before
    switch (e->GetType())
    {
    case UserEvent::EventType::Scale:
      {
        ref_ptr<ScaleEvent> scaleEvent = make_ref(e);
        breakAnim = OnSetScale(scaleEvent);
        TouchCancel(m_touches);
      }
      break;
// After
    switch (e->GetType())
    {
    case UserEvent::EventType::Scale:
    {
      ref_ptr<ScaleEvent> scaleEvent = make_ref(e);
      breakAnim = OnSetScale(scaleEvent);
      TouchCancel(m_touches);
    }
    break;
```

biodranik reviewed (approved):
Some examples look worse )

We are not forced to reformat all the code, right? We can start using it on the new or touched files.

AndrewShkrob commented:
What exactly looks worse? I can try to fix it

vng commented:
That's why I don't like auto formatters :) Probably, keep some basic checks.
Can we avoid this?
```
// After
std::string const & SymbolsTexture::SymbolKey::GetSymbolName() const { return m_symbolName; }
```

AndrewShkrob commented:
> That's why I don't like auto formatters

Why? Because you don't want to spend some time configuring them? 😄 

Yes, we can avoid this
```cpp
// After
std::string const & SymbolsTexture::SymbolKey::GetSymbolName() const { return m_symbolName; }
```

biodranik commented:
One-liners are not the biggest issue here (they save a lot of space and allow faster code reading).

vng commented:
We should set up some basic rules, but not _full coverage_.
It should not make my code one-line
It should not reformat the function's arguments - the author makes it better.
It should not reformat _columns_  code - the author makes it better.

Basic rules: spaces, braces, ok with switches, tabs, const & *.

biodranik commented:
@vng Can we force clang-format for newbies, and leave it as is for "experienced dinosaurs" ;-)?

vng commented:
@biodranik No, this is strange. If we can set up the rules I've described: spaces, braces, switches, tabs == 2 spaces, const & *, remove trailing whitespaces.
We can include it even in GA checks.

AndrewShkrob commented:
>It should not make my code one-line
It should not reformat the function's arguments - the author makes it better.
It should not reformat columns code - the author makes it better.

@vng I'm afraid you gave impossible restrictions for clang-format here
Can we discuss your requirements?

In general, the tool formats the code according to basic rules and guarantees that the code will have the same formatting in the same conditions. In some cases when we don't want to have the formatted code, we can use `// clang-format off`

>It should not make my code one-line

Why not? If the function has only one line and it fits the specified line length (100-120 symbols in our case), it will be formatted to one-line. We already have a lot of places with such one-line functions. And some places with functions that potentially can be formatted to one-line. Is it bad?
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#allowshortfunctionsonasingleline

>It should not reformat the function's arguments 

We can specify how the function's arguments will be formatted. But we have to choose one rule for all cases. That's the reason why we have a clang-format file and docs/CODESTYLE.md. There shouldn't be places where you decide that "this is how it's going to look better." One day you think one thing, and the next day another. We should formalize these places with the auto format. This is my pov.
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#alignafteropenbracket
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#binpackparameters
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#allowallparametersofdeclarationonnextline

>It should not reformat columns code

I don't know what is it about, but I'm sure everything I wrote above fits here as well

vng commented:
- I don't like one-line functions
- It's impossible to set a good breakpoint on a function's body, without a prolog

This is a simple ask - do not reformat arguments. Formatter should be a tool, not a restriction itself.

rtsisyk commented:
> * I don't like one-line functions

+1. I remember that at some point, IntelliJ started to collapse all one-liners into the one line it so super annoying that I stopped using it until they added an option... 

Can we disable this part for now and move forward? Also, formatting should be checked by CI. Otherwise nobody will do that.

biodranik reviewed (commented):
@AndrewShkrob let's revive it and use at least conditionally, when necessary, or for new files.

AndrewShkrob commented:
@organicmaps/cpp PTAL

I removed outdated `IncludeCategories`. To be honest, I don't see any sense in this configuration and don't want to fix it when it becomes obsolete again in a few months. 
I also removed parameters that are equal to default in the `Google` codestyle. Only the diff is left.

rtsisyk commented:
The problem with this PR is that the new automated tool reformats too much:

```
 2161 files changed, 63163 insertions(+), 71650 deletions(-)
```

Can we attempt to minimize the difference between the existing code style and the new style enforced by the tool as much as possible? 

A notable issue is placing the function value on a separate line:

```
 //  public static void nativeRemoveListener();
-JNIEXPORT void JNICALL
-Java_app_organicmaps_location_LocationState_nativeRemoveListener(JNIEnv * env, jclass clazz)
+JNIEXPORT void JNICALL Java_app_organicmaps_location_LocationState_nativeRemoveListener(JNIEnv * env, jclass clazz)
 {
```

Is this behaviour configurable somehow?

CC: @organicmaps/cpp, any thoughts?

vng commented:
@rtsisyk 
I have already ask a more generic question above. Is it possible not to change the formatting, which is not described in the rules?

We can set braces, spaces, switches, tabs, max line width. But I don't understand why _it wants_ to reformat column or manually aligned arguments code (like above).

AndrewShkrob commented:
> We can set braces, spaces, switches, tabs, max line width. But I don't understand why it wants to reformat column or manually aligned arguments code (like above).

This is a tool that works by a specified set of rules. And the purpose of this tool is to apply the same rules to every line in the whole codebase.

You want to see this code after reformat:
```cpp
void foo();
```

You don't want it to be reformatted this way:
```cpp
void
foo();
```

What's the difference here?
```cpp
 //  public static void nativeRemoveListener();
-JNIEXPORT void JNICALL
-Java_app_organicmaps_location_LocationState_nativeRemoveListener(JNIEnv * env, jclass clazz)
+JNIEXPORT void JNICALL Java_app_organicmaps_location_LocationState_nativeRemoveListener(JNIEnv * env, jclass clazz)
 {
```

Even though it is possible to apply specific rules to android/cpp folder but I don't see the reason why should we do it.

if you don't like this part...
```cpp
// Before
m2::RectF StipplePenPacker::MapTextureCoords(m2::RectU const & pixelRect) const
{
  return { (pixelRect.minX() + 0.5f) / m_canvasSize.x,
           (pixelRect.minY() + 0.5f) / m_canvasSize.y,
           (pixelRect.maxX() - 0.5f) / m_canvasSize.x,
           (pixelRect.maxY() - 0.5f) / m_canvasSize.y };
}
// After
m2::RectF StipplePenPacker::MapTextureCoords(m2::RectU const & pixelRect) const
{
  return {(pixelRect.minX() + 0.5f) / m_canvasSize.x, (pixelRect.minY() + 0.5f) / m_canvasSize.y,
          (pixelRect.maxX() - 0.5f) / m_canvasSize.x, (pixelRect.maxY() - 0.5f) / m_canvasSize.y};
}
```

Well, I feel you. That's a pity that clang-format don't have an explicit option for this (or I can't find it). But that's pretty obvious why it formats this line. Because the line width is configured to be 120.
What's the difference for the TOOL in these two segments of code?
```cpp
// 1
         { (pixelRect.minX() + 0.5f) / m_canvasSize.x,
           (pixelRect.minY() + 0.5f) / m_canvasSize.y,
           (pixelRect.maxX() - 0.5f) / m_canvasSize.x,
           (pixelRect.maxY() - 0.5f) / m_canvasSize.y }
// 2
{ (pixelRect.minX() + 0.5f) / m_canvasSize.x,
    (pixelRect.minY() + 0.5f) / m_canvasSize.y,
                  (pixelRect.maxX() - 0.5f) / m_canvasSize.x,
    (pixelRect.maxY() - 0.5f) / m_canvasSize.y }
```

I think it's a very small trade-off for the all good things this tool could potentially contribute to our project.

If there ever will be somewhere a place where you don't want the tool to reformat the code, you can always use `// clang-format on/off` comments.

And let's be honest, if I didn't give you this example, you would never notice these changes.
If you see such code in the PR
```cpp
m2::RectF StipplePenPacker::MapTextureCoords(m2::RectU const & pixelRect) const
{
  return { (pixelRect.minX() + 0.5f) / m_canvasSize.x,
           (pixelRect.minY() + 0.5f) / m_canvasSize.y,
           (pixelRect.maxX() - 0.5f) / m_canvasSize.x,
           (pixelRect.maxY() - 0.5f) / m_canvasSize.y };
}
```
You would say "Ok, nothing wrong with it"

If you see such code in the PR
```cpp
m2::RectF StipplePenPacker::MapTextureCoords(m2::RectU const & pixelRect) const
{
  return {(pixelRect.minX() + 0.5f) / m_canvasSize.x, (pixelRect.minY() + 0.5f) / m_canvasSize.y,
          (pixelRect.maxX() - 0.5f) / m_canvasSize.x, (pixelRect.maxY() - 0.5f) / m_canvasSize.y};
}
```
You would say "Ok, nothing wrong with it" as well

AndrewShkrob commented:
It's really annoying that we cannot agree on a codestyle for 2 years.
However, it was already agreed before that this project would use Google codestyle.
https://github.com/organicmaps/organicmaps/blob/master/docs/CPP_STYLE.md

pastk commented:
Obviously people have a different ideal format style in their minds. And obviously people will have to compromise to reach a consensus.

A not ideal automatic code formatting is better than the absence of it.
Let's avoid spending time in PRs applying formatting nits.

Let's move forward with this PR!

AndrewShkrob commented:
@vng I'd like to have your approve :)

AndrewShkrob commented:
Already merged
